### ISSUE TITLE
alter-var-root does not appear to change the variable's value

### ISSUE THREAD

**Issue by ikappaki** [basilisp-lang/basilisp#1166]
Hi,

I can't seem to get `alter-var-root` to alter any variable's root binding.

To reproduce, create a variable and alter its value. Even though the update happens, the variables root value remains the same as before the update
```clojure
> basilisp repl
basilisp.user=> (def abc 7)
#'basilisp.user/abc

basilisp.user=> (alter-var-root #'abc (fn [x] (println :current x) (inc x)))
:current 7
nil

basilisp.user=> abc
7
```

The same works in clojure as expected
```clojure
> clj
Clojure 1.12.0
user=> (def abc 7)
#'user/abc
user=> (alter-var-root #'abc (fn [x] (println :current x) (inc x)))
:current 7
8
user=>  abc
8
```

Thanks

**Comment by chrisrink10** [comment]
See #969

**Comment by chrisrink10** [comment]
I would accept a PR to change the docstring to include a warning about this behavior.

I would also consider a warning issued when calling `alter-var-root` on Vars not defined with `^:redef` or `^:dynamic`. I think trying to warn in every correct case may be prohibitively difficult since it is possible that not every namespace is compiled with the same compiler options.

**Comment by chrisrink10** [comment]
> Will do, I was also thinking of dropping a note in the differences with clojure section of the documentation.

Makes sense to me.

**Comment by ikappaki** [comment]
Hi @chrisrink10 , I'd also like to update `alter-var-root` to return the new value with this ticket, to make it the same with Clojure behavior please.

In Clojure, it returns the new value (even though it is [not documented](https://github.com/clojure/clojure/blob/clojure-1.11.1/src/clj/clojure/core.clj#L5530))
```clojure
> clj
Clojure 1.12.0
user=> (def abc 7)
#'user/abc
user=> (alter-var-root #'abc (fn [x] (println :prev x) (inc x)))
:prev 7
8
```

Basilisp currenty returns `nil`
```clojure
basilisp.user=> (def abc 7)
#'basilisp.user/abc

basilisp.user=> (alter-var-root #'abc (fn [x] (println :prev x) (inc x)))
:prev 7
nil
```

or let me know if you would like me to raise another issue for the same.

**Comment by chrisrink10** [comment]
Fixed in #1171
